### ISSUE TITLE
Handling targets in corners

### ISSUE THREAD

**Issue by gurupras** [pulsardev/vue-tour#23]
I ran into an issue while attempting to target elements at corners. The `.v-step__arrow` seems to always be in the middle of the tour step which, in turn, ends up pointing away from the actual target.

Here's a [pen](https://codepen.io/gurupras/pen/mLGagO) describing the issue. The example has 3 tour steps, the third of which highlights a different issue (the whole screen moves to accomodate the step and it doesn't move back), but I suspect this might just be an issue on my machine.

In short, this could probably be solved with `placement` values of `top-left`, `top-right` and so on. I strongly feel that this is something that's absolutely necessary for mobile since things tend to get pushed to corners a lot on small screens.

**Comment by mmorainville** [comment]
Hi @gurupras,

Thanks for bringing up this problem! Indeed the arrow's position is at the moment fixed in CSS.
This should be adaptable as we use Popper.js to handle the positionning at corners and they do it on their demo page: https://popper.js.org/

We'll take a look at it, but we won't be able to do that immediately. If you want to give it a try, a PR is welcome!

**Comment by mycarrysun** [comment]
@euxneks did you get this to work by adding the class? I am having the same problem - but looks like that class `popper_arrow` is not included in this package.

**Comment by euxneks** [comment]
Yeah, you'll have to download the source and modify the code to make that work - I am thinking of forking and making a PR but it's really like a 1 line change.

**Comment by euxneks** [comment]
@mycarrysun yeah you can add the css class `popper__arrow` to the arrow div element in `VStep.vue` in `/src/components` at line 25 here: https://github.com/pulsardev/vue-tour/blob/master/src/components/VStep.vue#L25 so it looks like this `class="popper__arrow v-step__arrow"` (I think that should be all that is needed)

**Comment by mycarrysun** [comment]
`popper__arrow` is not included in vue-tour.css. Where are you seeing this class?

**Comment by euxneks** [comment]
@mycarrysun That's the class that popper.js needs in order to know an element is an arrow, you can see the class in use for popper.js on their homepage. I didn't look into whether you can define the popper arrow class with popper.js but it's a quick, easy solution that solves the arrow adjustment problems.

**Comment by mycarrysun** [comment]
@euxneks I don't know if you are talking about something else but the styles that are applied on the page you are talking about are:
```css
/*this is for bottom placement as an example, other placements are similar*/
border-width: 0 5px 5px 5px;
border-left-color: transparent;
border-right-color: transparent;
border-top-color: transparent;
top: -5px;
left: calc(50% - 5px);
margin-top: 0;
margin-bottom: 0;

/*these are generic for all popper arrows*/
width: 0;
height: 0;
border-style: solid;
position: absolute;
margin: 5px;
```

This is the same styling that `.v-step__arrow` has which I'm sure @mmorainville just copied. 

There is going to need to be some javascript to move the arrow to a specific edge of the step box. CSS cannot calculate the position of the step in relation to the viewport.

**Comment by ex7r3me** [comment]
I tried `popper__arrow` which didn't work for me and I've found `x-arrow` property in the popper.js codes that let popper know which child element is the arrow. I've made a PR on it and I hope it helps

**Comment by mycarrysun** [comment]
Looks good @ex7r3me I think that will work. Although I think a better place to set this would be in [here](https://github.com/pulsardev/vue-tour/blob/9fbcfe5f702e4fe47e28b6c270fead0f07e1d554/src/shared/constants.js#L13)
Change it to:
```js
export const DEFAULT_STEP_OPTIONS = {
  placement: 'bottom',
  modifiers: {
    arrow: {
      element: '.v-step__arrow',
    }
  }
}
```

Also since the Popper.js api is exposed you can do this to achieve the same effect without the need to add #35 with each step
```js
let steps = [{
  target: '.step-1',
  content: 'Step 1 content',
  params:{
    modifiers:{
      arrow: {
        element: '.v-step__arrow'
      }
    }
  }
}]
```

**Comment by ex7r3me** [comment]
@mycarrysun that's cool. using popper API is a very good idean and I think the workaround can save someone's day.

**Comment by Outpox** [comment]
Looks good to me, may I reject #35 and merge #36 then?

**Comment by Outpox** [comment]
I've just merged #36, thanks for your work guys!  
I'll update the demo website + the doc to reflect this changes.

**Comment by mycarrysun** [comment]
@Outpox yes - thanks man!

**Comment by mycarrysun** [comment]
@Outpox um...so I have been making commits to my fork in my app and those commits also got merged. The only difference that technically could be a breaking change would be that when the target for the step is not found, the tour is not stopped - there is an event that is emitted instead so the user can handle what happens. If you guys think those changes should be reverted let me know and I'll create a new branch just for this issue (which I should have done in the first place).

**Comment by Outpox** [comment]
I'd say it's ok but it would need to be documented. I don't really remember how it was handled before, @mmorainville any tough on this?  
I'll leave it up to you if want to create a new issue or reopen this one ;-)

**Comment by mycarrysun** [comment]
The only thing I removed was that the VStep's `stop` method was called when the target was not found. This also triggered the `onStop` callback of course, so this does not happen either.

The event emitted is called `targetNotFound` and the payload that is sent is the step object that was not found. It is emitted by the VStep component - it would probably be better in the VTour component. I can make a PR for this (with it's own branch ;) )

**Comment by Outpox** [comment]
Thanks for remembering me :D  
Indeed I believe it would make more sense for the event to be emitted from the VTour component.  
BTW I'm quite busy this w-e and will only have the time to release a new build next week.

**Comment by mycarrysun** [comment]
@Outpox I am creating another option for if the user wants to stop the tour when the target is not found which will default to true. This will add the functionality back to the original way that this worked and allow me to use it the way I was. It should of course be documented and I can work on that next week but I don't have time right now.

PS - I also needed access to the popper element at one point so I added that to the `_data` object in the `VStep` component.

**Comment by tbredin** [comment]
Just checking — is this documented anywhere? I made it work by adding the `arrow` modifier:

```
let steps = [{
  target: '.step-1',
  content: 'Step 1 content',
  params:{
    modifiers:{
      arrow: {
        element: '.v-step__arrow'
      }
    }
  }
}]
```

But only found out how to do it after searching closed github issues for keyword "corner" as I couldn't see it in the docs anywhere. Apologies if I missed it

**Comment by tbredin** [comment]
Hmm, also — it doesn't appear to play well with the `placement` parameter
